### PR TITLE
Improvement/636 deploy calico

### PR DESCRIPTION
**Pull request by TeddyAndrieux**
Deploy calico-node

Fixes: #636

### CODE REVIEW

File: salt/_states/kubernetes.py [hunk 1]
```diff
@@ -1240,3 +1240,97 @@ def daemonset_present(
...
+            ret['result'] = None
+            ret['comment'] = 'The clusterrole is going to be created'
+            return ret
+
+        res = __salt__['kubernetes.create_clusterrole'](
+            name=name,
+            rules=rules,
+            **kwargs)
+        ret['result'] = True
+        ret['changes']['{0}'.format(name)] = {
```
Comment (NicolasT) [inline salt/_states/kubernetes.py]: Nit: this could be

```suggestion
        ret['changes'][name] = {
```

File: salt/_states/kubernetes.py [hunk 2]
```diff
@@ -1240,3 +1240,97 @@ def daemonset_present(
...
+        # TODO: improve checks  # pylint: disable=fixme
+        log.info('Forcing the recreation of the customresourcedefinition')
+        ret['comment'] = 'The custonresourcedefinition is already present. Forcing recreation'
+        res = __salt__['kubernetes.replace_customresourcedefinition'](
+            name=name,
+            spec=spec,
+            **kwargs)
+
+        ret['result'] = True
+        ret['changes']['{0}'.format(name)] = {
```
Comment (NicolasT) [inline salt/_states/kubernetes.py]:
```suggestion
        ret['changes'][name] = {
```

File: salt/metalk8s/calico/configured.sls [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        O: metalk8s:calico-node
+    - apiserver: https://{{ kube_api.service_ip }}:443
+    - cluster: {{ defaults.cluster }}
+
+Create CNI calico configuration file:
+  file.serialize:
+    - name: /etc/cni/net.d/10-calico.conflist
+    - user: root
+    - group: root
+    - mode: 664
```
Comment (NicolasT) [inline salt/metalk8s/calico/configured.sls]: `644`?

File: scripts/bootstrap.sh.in [hunk 4]
```diff
@@ -158,6 +158,11 @@ install_addons() {
     ${SALT_CALL} --retcode-passthrough state.apply metalk8s.kubeadm.init.addons saltenv=metalk8s-@VERSION@ pillarenv=metalk8s-@VERSION@
 }
 
+install_calico() {
+    ${SALT_CALL} --retcode-passthrough state.apply metalk8s.calico saltenv=metalk8s-@VERSION@ pillaren$
```
Comment (NicolasT) [inline scripts/bootstrap.sh.in]: Something wrong with `pillarenv`?

**Review by NicolasT — DISMISSED** [review]
Some minor issues.

File: salt/_modules/kubernetes.py [hunk 5]
```diff
@@ -1569,7 +1569,7 @@ def __dict_to_service_spec(spec):
         if key == 'ports':
             spec_obj.ports = []
             for port in value:
-                kube_port = kubernetes.client.V1ServicePort()
+                kube_port = kubernetes.client.V1ServicePort(port=port)
```
Comment (sayf-eddine-scality) [inline salt/_modules/kubernetes.py:1589]: this dirty fix should be removed and we shall apply the fix from the saltstack repo @NicolasT  ?

File: scripts/bootstrap.sh.in (hunk 4 above)
Comment (TeddyAndrieux) [reply scripts/bootstrap.sh.in]: Oh indeed

File: salt/metalk8s/calico/configured.sls (hunk 3 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/calico/configured.sls]: Calico default it to 664 :p but will change

File: salt/_modules/kubernetes.py (hunk 5 above)
Comment (NicolasT) [reply salt/_modules/kubernetes.py:1589]: Which fix would that be?

File: salt/_modules/kubernetes.py [hunk 6]
```diff
@@ -1975,3 +1975,178 @@ def replace_daemonset(
...
+            raise CommandExecutionError(exc)
+    finally:
+        _cleanup(**cfg)
+
+
+def create_customresourcedefinition(
+        name,
+        spec,
+        **kwargs):
+    # Workaround for https://github.com/kubernetes-client/python/issues/376
```
Comment (NicolasT) [inline salt/_modules/kubernetes.py]: You don't want to do this here, but once, at import time. The monkey-patch actually relies on this, the implementation as-is now would become more inefficient with every call to this procedure.

File: salt/_modules/kubernetes.py [hunk 7]
```diff
@@ -1975,3 +1975,178 @@ def replace_daemonset(
...
+                'Exception when calling '
+                'RbacAuthorizationV1Api->replace_cluster_role'
+            )
+            raise CommandExecutionError(exc)
+    finally:
+        _cleanup(**cfg)
+
+
+def show_customresourcedefinition(name, **kwargs):
+    cfg = _setup_conn(**kwargs)
```
Comment (NicolasT) [inline salt/_modules/kubernetes.py:2077]: I'm surprised this (and `replace_crd`) could succeed without the monkey-patch being in place. Anyway, see https://github.com/scality/metalk8s/pull/696/files#r261885992, must be placed elsewhere anyway.

Comment (TeddyAndrieux) [reply salt/_modules/kubernetes.py:2077]: `show` don't need the workaround for replace I got another issue will look at it but it's not the same

File: salt/_modules/kubernetes.py (hunk 5 above)
Comment (TeddyAndrieux) [reply salt/_modules/kubernetes.py:1589]: Just cherry picked all your commit @sayf-eddine-scality should I remove this one ?

Comment (sayf-eddine-scality) [reply salt/_modules/kubernetes.py:1589]: did not push any fix for this issue yet but I created a tracking ticket #697

Comment (sayf-eddine-scality) [reply salt/_modules/kubernetes.py:1589]: basically this fix will fail once we specify `target_port` or `source_port` in the state

File: salt/_modules/kubernetes.py [hunk 8]
```diff
@@ -1975,3 +1992,162 @@ def replace_daemonset(
...
         _cleanup(**cfg)
+
+
+def show_clusterrole(name, **kwargs):
+    cfg = _setup_conn(**kwargs)
+    try:
+        api_instance = kubernetes.client.RbacAuthorizationV1Api()
+        api_response = api_instance.read_cluster_role(name)
+
+        return api_response.to_dict()
```
Comment (charlesprost) [inline salt/_modules/kubernetes.py:2003]: If there is not evaluation/remote call in the `to_dict` call, it's probably better to put it into an `else` statement.

File: salt/_modules/kubernetes.py [hunk 9]
```diff
@@ -1975,3 +1992,162 @@ def replace_daemonset(
...
+        if isinstance(exc, ApiException) and exc.status == 404:
+            return None
+        else:
+            log.exception(
+                'Exception when calling '
+                'RbacAuthorizationV1Api->read_cluster_role'
+            )
+            raise CommandExecutionError(exc)
+    finally:
+        _cleanup(**cfg)
```
Comment (charlesprost) [inline salt/_modules/kubernetes.py:2014]: You can definitely mutualize this whole try/except/finally pattern between the different methods

File: salt/metalk8s/calico/configured.sls [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+Create CNI calico configuration file:
+  file.serialize:
+    - name: /etc/cni/net.d/10-calico.conflist
+    - user: root
+    - group: root
+    - mode: 644
+    - makedirs: True
+    - dir_mode: 755
+    - formatter: json
+    - dataset:
```
Comment (charlesprost) [inline salt/metalk8s/calico/configured.sls:28]: Could be interested to have the source or the why for values in this dataset

File: salt/metalk8s/calico/configured.sls [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+          - type: "portmap"
+            snat: true
+            capabilities:
+              portMappings: true
+    - require:
+      - metalk8s_kubeconfig: Create kubeconf file for calico
+
+{%- else %}
+
+Unable to generate calico kubeconf file, no CA Server available:
```
Comment (charlesprost) [inline salt/metalk8s/calico/configured.sls]:
```suggestion
Unable to generate calico kubeconf file, no CA server available:
```

File: salt/metalk8s/calico/deployed.sls [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+        group: crd.projectcalico.org
+        version: v1
+        names:
+          kind: BGPPeer
+          plural: bgppeers
+          singular: bgppeer
+    - require_in:
+      - kubernetes: Deploy calico-node (ClusterRole)
+
+Deploy bgpconfigurations.crd.proejctcalico.org (CustomResourceDefinition):
```
Comment (charlesprost) [inline salt/metalk8s/calico/deployed.sls]:
```suggestion
Deploy bgpconfigurations.crd.projectcalico.org (CustomResourceDefinition):
```

File: salt/metalk8s/calico/deployed.sls [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+        group: crd.projectcalico.org
+        version: v1
+        names:
+          kind: BGPConfiguration
+          plural: bgpconfigurations
+          singular: bgpconfiguration
+    - require_in:
+      - kubernetes: Deploy calico-node (ClusterRole)
+
+Deploy ippools.crd.proejctcalico.org (CustomResourceDefinition):
```
Comment (charlesprost) [inline salt/metalk8s/calico/deployed.sls]:
```suggestion
Deploy ippools.crd.projectcalico.org (CustomResourceDefinition):
```

File: salt/metalk8s/calico/deployed.sls [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+        group: crd.projectcalico.org
+        version: v1
+        names:
+          kind: IPPool
+          plural: ippools
+          singular: ippool
+    - require_in:
+      - kubernetes: Deploy calico-node (ClusterRole)
+
+Deploy hostendpoints.crd.proejctcalico.org (CustomResourceDefinition):
```
Comment (charlesprost) [inline salt/metalk8s/calico/deployed.sls]:
```suggestion
Deploy hostendpoints.crd.projectcalico.org (CustomResourceDefinition):
```

**Review by charlesprost — DISMISSED** [review]

File: salt/_modules/kubernetes.py (hunk 9 above)
Comment (TeddyAndrieux) [reply salt/_modules/kubernetes.py:2014]: Yes it's something we need to do.
Btw this file and the one in `_states` directory it's just a copy from the `salt` repo with our new function, which is, for the moment, copy/paste for new kubernetes stuff (maybe in futur version of salt we will remove this file + the `_states/kubernete.py`).
So the structure is the same as others salt function.

File: salt/_modules/kubernetes.py (hunk 8 above)
Comment (TeddyAndrieux) [reply salt/_modules/kubernetes.py:2003]: Yes we just copy/pasted from salt module for the moment. But yes indeed

File: salt/metalk8s/calico/configured.sls (hunk 10 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/calico/configured.sls:28]: Hmm yes it's basically a CNI configuration file but yes we can

File: salt/metalk8s/calico/deployed.sls (hunk 13 above)
Comment (TeddyAndrieux) [reply salt/metalk8s/calico/deployed.sls]: Wow good catch

**Review by charlesprost — APPROVED** [review]

**Comment by TeddyAndrieux** [comment]
@bert-e approve

**Pull request merged**